src/imfilter.jl

```diff
@@ -6,18 +6,31 @@ function imfilter_cpu(img, krn)
   imfilter(img, centered(krn), Inner(), Algorithm.FFT())
 end
 
+# preallocated memory for krn optimization on GPU
+krnbuffer_gpu = nothing
+
+function alloc_krn_buffer_gpu!(sz)
+  global krnbuffer_gpu
+  
+  if isnothing(krnbuffer_gpu) || size(krnbuffer_gpu) != sz
+    krnbuffer_gpu = CuArray(zeros(sz))
+  end
+
+  krnbuffer_gpu
+end
+
 function imfilter_gpu(img, krn)
   # retrieve basic info
   N = ndims(img)
   T = eltype(img)
 
-  # pad kernel to common size with image
-  padsize = size(img) .- size(krn)
-  padkrn  = padarray(krn, Fill(zero(T), ntuple(i->0, N), padsize))
+  # copy krn data to preallocated GPU memory
+  paddedkrn = alloc_krn_buffer_gpu!(size(img))
```

Comment:
This is not the right place to call this allocation right? Why not call it in the main iqsim function before the loop? Calling it inside here seems counter intuitive, at least to me reading the code for the first time.

Reply:
refactored